### ISSUE TITLE
Allow creating foreign keys in sqlite

### ISSUE THREAD

**Issue by B4nan** [knex/knex#3351]
# Environment

Knex version: 0.18.3
Database + version: sqlite3 v4.0.7
OS: macos

# Bug / Feature request

Currently knex complaints about adding a foreign key when altering a table (`SQLite3 Foreign & Primary keys may only be added on create`). I understand why you have it there, but it is not 100% true. You can create foreign keys on existing table - you just have to add new column with that key, the thing you cannot do is to alter existing column (making it foreign key). 

Currently I am trying to use knex as a query builder/runner in my ORM (https://github.com/B4nan/mikro-orm), and the way I am talking about is how it works there currently (so it is not theoretical, it is something I already have implemented without knex - see here https://github.com/B4nan/mikro-orm/tree/master/lib/schema). 

Here is example SQL that is valid and I would like to be able to build that with knex schema builder:

```sql
ALTER TABLE "author" ADD "favourite_book_id" INTEGER DEFAULT NULL REFERENCES "book" ("id") ON DELETE SET NULL ON UPDATE CASCADE;
```

The reason why I don't want to create this FK when creating the table is that there are references in both ways - author table references book and vice versa. It is a model example, not a real world usage.

**Comment by rijkvanzanten** [comment]
@kibertoad Any updates on this? I noticed you assigned yourself so I don't want to pick this up if you're already working on it 🙂

**Comment by rijkvanzanten** [comment]
I'm imagining that implementing this for SQLite is effectively a copy of https://github.com/knex/knex/pull/4092 but instead of removing `REFERENCES etc etc` we add it?

**Comment by kibertoad** [comment]
@rijkvanzanten Not working on it right now, so if you have some spare time, would appreciate a contribution!

**Comment by kibertoad** [comment]
@B4nan Apologies for this taking waaaaay longer than expected, it wasn't the simplest feature to implement, but thanks to @rijkvanzanten it is finally there. Now it's available in Knex 0.21.15. Could you please confirm that it works as expected for your use-case?

**Comment by B4nan** [comment]
Unfortuately this isn't working as I would expect (or better to say, it is not working at all for my use case) - knex is probably trying to be too smart. I am using knex as a QB, so I want it to give me queries needed for the job. Instead, I see things like `PRAGMA table_info(`table_name`);` and no FKs are actually added. 

I am using the `SchemaBuilder.toQuery()` method to obtain the queries, and the `PRAGMA table_info` is among them (while no FK query is present). Here is a snippet of what I am doing:

```ts
const builder = knex.schema.alterTable('book', table => {
  const col = table.specificType('author_id', 'integer');
  col.references('id').inTable('author');
});
const queries = builder.toQuery();
```

And it gives me:

```sql
alter table `book` add column `author_id` integer;
PRAGMA table_info(`book`)
```

**Comment by kibertoad** [comment]
@B4nan Not sure if it's even possible to generate a standalone SQL query here, since from my understanding what the implementation does is basically rewrite the table DDL definition inside the DB, so there is no incremental statement being applied at all. @rijkvanzanten Can you provide additional comments?

**Comment by kibertoad** [comment]
@B4nan Does SQLite even support incremental introduction of a foreign key via an ALTER TABLE statement?

**Comment by B4nan** [comment]
Already mentioned in the OP, that I am after something I had implementedd myself in MikroORM without knex. So yes, it is possible, the query in the OP is valid and functional, and this issue is about not being able to create such query via knex.

So it sounds like a confirmation of what I already understood some time ago. Knex is not really a query builder :]

**Comment by kibertoad** [comment]
@B4nan Ideally it strives to be a query builder, and your point is valid, and since it is doable (apologies, didn't double-check your original issue description before writing last comment), we should amend our implementation to build a query instead. I'll give it a shot.

**Comment by kibertoad** [comment]
What current implementation covers is adding foreign key to an **existing** column, and that doesn't seem to be possible without hacking the DDL. However, for the scenario that you depicted, when we are adding a brand new column, it should be possible to do it in one go.

**Comment by B4nan** [comment]
In the end, everything is actually a query. So I'd argue that it is not possible to do things knex is doing, while still being a QB first. It is fine to use `PRAGMA table_info` behind the scenes, but that should be implementation detail, not part of the output of `toQuery()` method.

**Comment by kibertoad** [comment]
@B4nan You are right. We should also adjust existing implementation to emit corrent `toQuery()` for what it is already doing as well.

**Comment by B4nan** [comment]
That would be great, another example of this problem is renaming colums. I remember I had to implement that from scratch due to knex giving me something like `PRAGMA table_info` there too, instead of actual queries.

**Comment by kibertoad** [comment]
@B4nan Started adapting code to support your use-case, but there is a nuance that makes it tricky - `toQuery` is a synchronous operation, while quite a lot of logic inside SQLite manipulations are being done asynchronously (e. g. we extract all existing rows to prepare insert query, retrieve current table definition for further modification, etc). Rewriting SQLs to do everything on-the-fly without any interim JS code would be extremely tricky and time-consuming.
Would it be OK for you if we introduced a new method, e. g. `builder.generateDdlQuery()` that would be async and would execute some read-only DB operations behind-the-scenes?

**Comment by B4nan** [comment]
Sounds good

**Comment by B4nan** [comment]
Actually this is still not working properly. While #4194 is great for advanced sqlite migrations, here it was never needed - in the OP I was asking for support to add new column with a FK - which is possible with sqlite without doing the temp table magic. Currently knex will first add the new column as one query, and then tries to do the temp table magic - but the new column is not yet added in the information schema and therefore missing in the following query most probably.

```sql
alter table `user` add column `profile_id` text null;
CREATE TABLE `_knex_temp_alter644` (`id` text not null, primary key (`id`), CONSTRAINT user_profile_id_foreign FOREIGN KEY (`profile_id`)  REFERENCES `profile` (`id`) ON UPDATE cascade ON DELETE cascade);
INSERT INTO _knex_temp_alter644 SELECT * FROM user;;
DROP TABLE "user";
ALTER TABLE "_knex_temp_alter644" RENAME TO "user";
```

Which obviously fails as the temp table references `profile_id` column in the FK, but it is not listed in its columns. The repro I [provided earlier](https://github.com/knex/knex/issues/3351#issuecomment-751484733) is still valid and produces exactly this issue.

What I was asking for was just one query:

```sql
alter table `user` add column `profile_id` text null CONSTRAINT user_profile_id_foreign REFERENCES `profile` (`id`) ON UPDATE cascade ON DELETE cascade;
```

(btw would be great if the queries are all lowercase, not sure why something is upper case and something not, or ideally have a way to control this, but keep in consistent in the first place - even in a single query sometimes the casing is not matching)

**Comment by B4nan** [comment]
Can we reopen this or should I create new issue? As mentioned above, this is still not fixed at all, the original issue in the OP is not addressed, and the new way via `generateDdlQuery` just fails in a different way (as the create temp table query ignores the just added column).

I was able to get around this for simple columns by patching knex again, but it is rather nasty hack than a real solution (which should be to respect the in memory state when creating the create temp table statement).

cc @kibertoad

**Comment by SephReed** [comment]
I found this fix on stack overflow:  https://stackoverflow.com/questions/65423969/foreign-key-constraint-ignored-by-knex-js

Basically, all you have to do is run this:  
```
knex.raw("PRAGMA foreign_keys = ON;").then(() => {
    console.log("Foreign Key Check activated.");
});
```

I see no reason why this should not be enabled by default.

**Comment by mercmobily** [comment]
I have triaged this as an urgent but. I realise that MikroORM no longer uses Knex.
